lib/buildkite/test_collector/tracer.rb

```diff
@@ -35,21 +35,30 @@ def initialize(section, start_at, end_at, detail)
         @children = []
       end
 
+      def duration
+        raise IncompleteSpan if end_at.nil?
```

Comment:
I can't see this error being caught anywhere? If that's the case, I think this will might cause Rspec to fail with that "error outside of execution" error message and/or fail the job?

Reply:
Hmm yeah.

I think this is what Java would call an `IllegalStateException`, and is basically a runtime assertion that some internal usage hasn't gone wrong; calling `duration` before setting `end_at` makes no sense, and as far as we can tell nothing is or should be doing that.

My normal default is to crash with a clear error message when something “impossible” happens.

But perhaps I need to rethink that in test collectors where it's more important that the show goes on, even if that means the test collector can't behave sanely. 🤔

I'm pretty sure without this check, the old code would have done this instead:

```
NoMethodError: undefined method `-' for nil:NilClass
```

So I don't think it's a regression.

Reply:
Oh right I see, you're setting the `end_at` right before the `duration` method is called - so the failure case would be if the call to Monotonic time library fails, which "should be" impossible. Is it worth putting in such a check for a dependant library? Not sure why the time call would fail? And seeing as this would have raised an exception in the past anyways, and we've never seen a case of that with the collector, maybe its fine to just rely on monotonic time doing its thing and not having an explicit check for it?

Reply:
bump on this comment :)

Reply:
Yeah I think it's fine; I don't think the monotonic time check can realistically return `nil`, if anything it might raise an error and crash the whole party.

So yeah I think it's fine like this.